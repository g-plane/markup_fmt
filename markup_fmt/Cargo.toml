[package]
name = "markup_fmt"
version = "0.10.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "Configurable HTML/Vue/Svelte/Astro/Angular/Jinja/Twig/Nunjucks/Vento formatter."
repository = "https://github.com/g-plane/markup_fmt"
license = "MIT"
exclude = ["/tests"]

[dependencies]
aho-corasick = "1.1"
css_dataset = "0.3"
itertools = "0.13"
memchr = "2.7"
once_cell = "1.19"
serde = { version = "1.0", optional = true }
tiny_pretty = { version = "0.2", features = ["unicode-width"] }

[dev-dependencies]
insta = { version = "1.39", features = ["glob"] }
similar-asserts = "1.5"
toml = "0.8"

[features]
config_serde = ["serde"]
