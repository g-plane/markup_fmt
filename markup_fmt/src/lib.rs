#![doc = include_str!("../README.md")]

mod ast;
pub mod config;
mod ctx;
mod error;
mod helpers;
mod parser;
mod printer;
mod state;

use crate::{config::FormatOptions, ctx::Ctx, parser::Parser, printer::DocGen, state::State};
pub use crate::{error::*, parser::Language};
use std::{borrow::Cow, path::Path};
use tiny_pretty::{IndentKind, PrintOptions};

/// Format the given source code.
///
/// An external formatter is required for formatting code
/// inside `<script>` or `<style>` tag.
/// If you don't need to format them or you don't have available formatters,
/// you can pass a closure that returns the original code. (see example below)
///
/// ```
/// use markup_fmt::{format_text, Language};
///
/// let code = r#"
/// <html>
///    <head>
///      <title>Example</title>
///      <style>button { outline: none; }</style>
///   </head>
///   <body><script>const a = 1;</script></body>
/// </html>"#;
///
/// let formatted = format_text(
///     code,
///     Language::Html,
///     &Default::default(),
///     |_, code, _| Ok::<_, std::convert::Infallible>(code.into()),
/// ).unwrap();
/// ```
///
/// For the external formatter closure,
///
/// - The first argument is file path.
/// Either script code or style code will be passed to this closure,
/// so you need to check file extension with the file path.
/// - The second argument is code that needs formatting.
/// - The third argument is print width that you may need to pass to formatter
/// if they support such option.
pub fn format_text<E, F>(
    code: &str,
    language: Language,
    options: &FormatOptions,
    external_formatter: F,
) -> Result<String, FormatError<E>>
where
    F: for<'a> FnMut(&Path, &'a str, usize) -> Result<Cow<'a, str>, E>,
{
    let mut parser = Parser::new(code, language.clone());
    let ast = parser.parse_root().map_err(FormatError::Syntax)?;
    let mut ctx = Ctx {
        language,
        indent_width: options.layout.indent_width,
        print_width: options.layout.print_width,
        options: &options.language,
        indent_level: 0,
        external_formatter,
        external_formatter_errors: Default::default(),
    };

    let doc = ast.doc(
        &mut ctx,
        &State {
            current_tag_name: None,
            is_root: true,
            in_svg: false,
        },
    );
    if !ctx.external_formatter_errors.is_empty() {
        return Err(FormatError::External(ctx.external_formatter_errors));
    }

    Ok(tiny_pretty::print(
        &doc,
        &PrintOptions {
            indent_kind: if options.layout.use_tabs {
                IndentKind::Tab
            } else {
                IndentKind::Space
            },
            line_break: options.layout.line_break.clone().into(),
            width: options.layout.print_width,
            tab_size: options.layout.indent_width,
        },
    ))
}

/// Detect language from file extension.
pub fn detect_language(path: impl AsRef<Path>) -> Option<Language> {
    let path = path.as_ref();
    match path.extension().and_then(std::ffi::OsStr::to_str) {
        Some("html") => {
            if path
                .file_stem()
                .map(|file_stem| file_stem.to_string_lossy().ends_with(".component"))
                .unwrap_or_default()
            {
                Some(Language::Angular)
            } else {
                Some(Language::Html)
            }
        }
        Some("vue") => Some(Language::Vue),
        Some("svelte") => Some(Language::Svelte),
        Some("astro") => Some(Language::Astro),
        Some("jinja" | "jinja2" | "twig" | "njk") => Some(Language::Jinja),
        Some("vto") => Some(Language::Vento),
        _ => None,
    }
}
