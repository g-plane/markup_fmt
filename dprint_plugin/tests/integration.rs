use anyhow::Error;
use insta::{Settings, assert_snapshot, glob};
use markup_fmt::{
    FormatError,
    config::{FormatOptions, ScriptFormatter},
    detect_language, format_text,
};
use std::{borrow::Cow, fs, io, path::Path};

#[test]
fn integration_with_dprint_ts_snapshot() {
    fn format_with_dprint_ts(input: &str, path: &Path) -> Result<String, FormatError<Error>> {
        let mut options = match fs::read_to_string(path.with_extension("toml")) {
            Ok(file) => toml::from_str::<FormatOptions>(&file).unwrap(),
            Err(e) if e.kind() == io::ErrorKind::NotFound => Default::default(),
            Err(e) => panic!("{e}"),
        };
        let file_name = path.file_name().and_then(|file_name| file_name.to_str());
        if file_name.is_some_and(|file_name| file_name.starts_with("deno")) {
            options.language.script_indent = true;
            options.language.style_indent = true;
        }
        options.language.script_formatter = Some(ScriptFormatter::Dprint);
        format_text(
            input,
            detect_language(path).unwrap(),
            &options,
            |code, hints| -> anyhow::Result<Cow<str>> {
                let ext = hints.ext;
                let additional_config =
                    dprint_plugin_markup::build_additional_config(hints, &options);
                if let Some(syntax) = malva::detect_syntax(&Path::new("file").with_extension(ext)) {
                    malva::format_text(
                        code,
                        syntax,
                        &serde_json::to_value(additional_config)
                            .and_then(serde_json::from_value)?,
                    )
                    .map(Cow::from)
                    .map_err(Error::from)
                } else if ext == "json" {
                    dprint_plugin_json::format_text(
                        &Path::new("file").with_extension(ext),
                        code,
                        &dprint_plugin_json::configuration::resolve_config(
                            additional_config,
                            &Default::default(),
                        )
                        .config,
                    )
                    .map(|formatted| {
                        if let Some(formatted) = formatted {
                            Cow::from(formatted)
                        } else {
                            Cow::from(code)
                        }
                    })
                } else {
                    dprint_plugin_typescript::format_text(
                        dprint_plugin_typescript::FormatTextOptions {
                            path: &Path::new("file").with_extension(ext),
                            extension: Some(ext),
                            text: code.to_owned(),
                            config: &dprint_plugin_typescript::configuration::resolve_config(
                                additional_config,
                                &Default::default(),
                            )
                            .config,
                            external_formatter: None,
                        },
                    )
                    .map(|formatted| {
                        if let Some(formatted) = formatted {
                            Cow::from(formatted)
                        } else {
                            Cow::from(code)
                        }
                    })
                }
            },
        )
    }

    glob!(
        "integration/**/*.{html,vue,svelte,astro,jinja,njk,vto}",
        |path| {
            let input = fs::read_to_string(path).unwrap();
            let output = format_with_dprint_ts(&input, path)
                .map_err(|err| format!("failed to format '{}': {:?}", path.display(), err))
                .unwrap();

            assert!(
                output.ends_with('\n'),
                "formatted output should contain trailing newline: {}",
                path.display()
            );

            let regression_format = format_with_dprint_ts(&output, path)
                .map_err(|err| {
                    format!(
                        "syntax error in stability test '{}': {:?}",
                        path.display(),
                        err
                    )
                })
                .unwrap();
            similar_asserts::assert_eq!(
                output,
                regression_format,
                "'{}' format is unstable",
                path.display()
            );

            build_settings(path.parent().unwrap().join("dprint_ts")).bind(|| {
                let name = path.file_name().unwrap().to_str().unwrap();
                assert_snapshot!(name, output);
            });
        }
    );
}

#[test]
fn integration_with_biome_snapshot() {
    fn format_with_biome(input: &str, path: &Path) -> Result<String, FormatError<Error>> {
        let mut options = match fs::read_to_string(path.with_extension("toml")) {
            Ok(file) => toml::from_str::<FormatOptions>(&file).unwrap(),
            Err(e) if e.kind() == io::ErrorKind::NotFound => Default::default(),
            Err(e) => panic!("{e}"),
        };
        options.language.script_formatter = Some(ScriptFormatter::Biome);
        format_text(
            &input,
            detect_language(path).unwrap(),
            &options,
            |code, hints| -> anyhow::Result<Cow<str>> {
                let ext = hints.ext;
                let mut additional_config =
                    dprint_plugin_markup::build_additional_config(hints, &options);
                additional_config.insert("javascriptIndentStyle".into(), "space".into());
                if let Some(syntax) = malva::detect_syntax(&Path::new("file").with_extension(ext)) {
                    malva::format_text(
                        code,
                        syntax,
                        &serde_json::to_value(additional_config)
                            .and_then(serde_json::from_value)?,
                    )
                    .map(Cow::from)
                    .map_err(Error::from)
                } else {
                    dprint_plugin_biome::format_text(
                        &Path::new("file").with_extension(ext),
                        code,
                        &serde_json::to_value(additional_config)
                            .and_then(serde_json::from_value)
                            .unwrap_or_default(),
                    )
                    .map(|formatted| {
                        if let Some(formatted) = formatted {
                            Cow::from(formatted)
                        } else {
                            Cow::from(code)
                        }
                    })
                }
            },
        )
    }

    glob!(
        "integration/**/*.{html,vue,svelte,astro,jinja,njk,vto}",
        |path| {
            let file_name = path.file_name().and_then(|file_name| file_name.to_str());
            if file_name.is_some_and(|file_name| {
                file_name == "return.astro" || file_name.starts_with("deno")
            }) {
                // dprint-plugin-biome doesn't support top-level return (but Biome supports),
                // so skip it.
                // Also, skip tests for Deno only.
                return;
            }

            let input = fs::read_to_string(path).unwrap();
            let output = format_with_biome(&input, path)
                .map_err(|err| format!("failed to format '{}': {:?}", path.display(), err))
                .unwrap();

            assert!(
                output.ends_with('\n'),
                "formatted output should contain trailing newline: {}",
                path.display()
            );

            let regression_format = format_with_biome(&output, path)
                .map_err(|err| {
                    format!(
                        "syntax error in stability test '{}': {:?}",
                        path.display(),
                        err
                    )
                })
                .unwrap();
            similar_asserts::assert_eq!(
                output,
                regression_format,
                "'{}' format is unstable",
                path.display()
            );

            build_settings(path.parent().unwrap().join("biome")).bind(|| {
                let name = path.file_name().unwrap().to_str().unwrap();
                assert_snapshot!(name, output);
            });
        }
    );
}

fn build_settings(path: impl AsRef<Path>) -> Settings {
    let mut settings = Settings::clone_current();
    settings.set_snapshot_path(path);
    settings.remove_snapshot_suffix();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_input_file();
    settings.set_omit_expression(true);
    settings.remove_input_file();
    settings.remove_info();
    settings
}
