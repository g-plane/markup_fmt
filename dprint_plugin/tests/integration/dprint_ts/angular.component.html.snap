---
source: dprint_plugin/tests/integration.rs
---
<button type="button" (click)="deleteHero()">Delete hero</button>

<tr><td [colSpan]="1 + 1">Three-Four</td></tr>

<p>The hero's birthday is {{ birthday | date }}</p>
